pinot-controller/src/main/java/org/apache/pinot/controller/cursors/ResponseStoreCleaner.java

```diff
@@ -0,0 +1,222 @@
+    String responseStoreCleanerTaskInitialDelay =
+        config.getProperty(CommonConstants.CursorConfigs.RESPONSE_STORE_CLEANER_INITIAL_DELAY);
+    if (responseStoreCleanerTaskInitialDelay != null) {
+      initialDelay = TimeUnit.SECONDS.convert(TimeUtils.convertPeriodToMillis(responseStoreCleanerTaskInitialDelay),
+          TimeUnit.MILLISECONDS);
+    }
+    return initialDelay;
+  }
+
+  private static long getFrequencyInSeconds(ControllerConf config) {
+    long frequencyInSeconds = TimeUnit.SECONDS.convert(
+        TimeUtils.convertPeriodToMillis(CommonConstants.CursorConfigs.DEFAULT_RESPONSE_STORE_CLEANER_FREQUENCY_PERIOD),
+        TimeUnit.MILLISECONDS);
+    String responseStoreCleanerTaskPeriod =
+        config.getProperty(CommonConstants.CursorConfigs.RESPONSE_STORE_CLEANER_FREQUENCY_PERIOD);
+    if (responseStoreCleanerTaskPeriod != null) {
+      frequencyInSeconds = TimeUnit.SECONDS.convert(TimeUtils.convertPeriodToMillis(responseStoreCleanerTaskPeriod),
+          TimeUnit.MILLISECONDS);
+    }
+
+    return frequencyInSeconds;
+  }
+
+  @Override
+  protected void processTables(List<String> tableNamesWithType, Properties periodicTaskProperties) {
+    long cleanAtMs = System.currentTimeMillis();
+    String cleanAtMsStr = periodicTaskProperties.getProperty(CLEAN_AT_TIME);
+    if (cleanAtMsStr != null) {
+      cleanAtMs = Long.parseLong(cleanAtMsStr);
+    }
+    doClean(cleanAtMs);
+  }
+
+  public void doClean(long currentTime) {
+    List<InstanceConfig> brokerList = _pinotHelixResourceManager.getAllBrokerInstanceConfigs();
+    Map<String, InstanceInfo> brokers = new HashMap<>();
+    for (InstanceConfig broker : brokerList) {
+      brokers.put(getInstanceKey(broker.getHostName(), broker.getPort()),
+          new InstanceInfo(broker.getInstanceName(), broker.getHostName(), Integer.parseInt(broker.getPort())));
+    }
+
+    try {
+      Map<String, String> requestHeaders = AuthProviderUtils.makeAuthHeadersMap(_authProvider);
+
+      Map<String, List<CursorResponseNative>> brokerCursorsMap = getAllQueryResults(brokers, requestHeaders);
+
+      String protocol = _controllerConf.getControllerBrokerProtocol();
+      int portOverride = _controllerConf.getControllerBrokerPortOverride();
+
+      List<String> brokerUrls = new ArrayList<>();
+      for (Map.Entry<String, List<CursorResponseNative>> entry : brokerCursorsMap.entrySet()) {
+        for (CursorResponse response : entry.getValue()) {
+          if (response.getExpirationTimeMs() <= currentTime) {
+            InstanceInfo broker = brokers.get(entry.getKey());
+            int port = portOverride > 0 ? portOverride : broker.getPort();
+            brokerUrls.add(
+                String.format(DELETE_QUERY_RESULT, protocol, broker.getHost(), port, response.getRequestId()));
+          }
+        }
+        Map<String, String> deleteStatus = getResponseMap(requestHeaders, brokerUrls, "DELETE", HttpDelete::new);
```

Comment:
Isn't this going to result in duplicate deletes for response stores that are not broker specific (blob stores etc.)? I see that you've documented this issue in the `ResponseStore` interface warning about multiple deletes and even potentially concurrent deletes. I don't get why that's required though? Can't we deduplicate based on the request ID from the cursor response and only issue one delete request per query (to any one broker) here?

Reply:
Now `brokerId` is initialized in the `ResponseStore`. It will return a response only if the `BrokerResponse.getBrokerId().equals(_brokerId)` - even if it has access to other responses. This will solve the duplicate entries in a list of response. This was easier to solve than I expected.

Reply:
I've also updated java docs etc.

Reply:
> It will return a response only if the BrokerResponse.getBrokerId().equals(_brokerId) - even if it has access to other responses. This will solve the duplicate entries in a list of response

Why do we want to do that? If a user is using S3 as the response store, don't we want any broker to be able to show and access the stored responses?

Reply:
I dont think its right for behavior/API results to magically change based on which FS is used. There should be explicit contracts for shared storage and shared nothing response stores. A response store should be tagged as one type or the other so that the base infra can use it the right way. e.g. `ResponseStoreCleaner` calls `getAllResponseStore` to get owned response stores. Another fn `getAllAccessibleResponseStores` to get all of them. I am thinking about the right way to make the diff clear through the interface.

Reply:
i had assumed that shared storage is superior. However in production it hasnt been that useful. In a multi-broker setup, clients will connect to a broker and contacting the same broker for paging through is fine. Even if there is a load balancer/proxy, the response has the broker host/port to request the proxy to connect to a specific broker. The pinning mechanism may differ based on the proxy. So I am leaning towards a shared nothing model (even if underneath it is a shared storage).

Reply:
Resolving this based on our discussion offline. While it may not always be possible for users to configure proxy / LB routing to specific brokers depending on their setup, we plan to introduce a broker <-> broker redirection model in the future which should help this use case as well as others. We're moving forward with the broker <-> response store association model.